lib/db.js

```diff
@@ -909,6 +909,24 @@ module.exports = (
       )
   }
 
+  DB.prototype.createSigninCode = function (uid) {
+    log.trace({ op: 'DB.createSigninCode' })
+
+    return random(config.signinCodeSize)
+      .then(code => {
+        const data = { uid, createdAt: Date.now() }
+        return this.pool.put(`/signinCodes/${code.toString('hex')}`, data)
+          .then(() => code, err => {
+            if (isRecordAlreadyExistsError(err)) {
+              log.warn({ op: 'DB.createSigninCode.duplicate' })
+              return this.createSigninCode(uid)
```

Comment:
One collision is super-unlikely, two collisions is super-unlikely squared and three collisions is super-unlikely cubed. I don't really see the need for a hard limit myself, but I'll add one if someone can tell me what the correct arbitrary value is.

Although also, I'll point out that a hard limit of sorts does exist when the process runs out of stack space (not that we'd ever get remotely that far of course).